Extend tests and automation scripts for airgapped installations (#682)
canonical/bundle-kubeflow: scripts/airgapped/README.md

Author: @DnPlas @ca-scribner do you think a graph in the README like the one I commented bellow https://github.com/canonical/bundle-kubeflow/pull/682#issuecomment-1829570436 could help?

While it touches a bit on the scripts, it's more aimed on the high level picture

# Airgap Utility Scripts

This directory contains bash and python scripts that are useful for performing
an airgapped installation. These scripts could either be used independently
to create airgap artifacts or via our testing scripts.

We'll document some use-case scenarios here for the different scripts.

## Prerequisites

To use the scripts in this directory you'll need to install a couple of Python
and Ubuntu packages.
```
pip3 install -r requirements.txt
sudo apt install pigz
```

## Get list of all images in bundle

In a lot of cases we need to have the list of all OCI Images used by a bundle.

> To get this list we can use the following script. This is scripts makes the

Reviewer: ```suggestion
To get this list we can use the following script. This script makes the
```

> In a lot of cases we need to have the list of all OCI Images used by a bundle.
> To get this list we can use the following script. This is scripts makes the

Reviewer: Not a mandatory change, but I think you just need to explain script usage. 
 You don't need to convince them it was important to make the script

```suggestion
Use `./scripts/airgapped/get-all-images.sh` to get the list of all OCI Images used by a bundle. This script makes the
```

Author: Added a commit for this, since the discussion was outdated due to other commits
https://github.com/canonical/bundle-kubeflow/pull/682/commits/afc70a9f5107ac25444c8bdba393def5640d41b2

following assumptions:
1. Every charm in the bundle has a `_github_repo_name` metadata field,
   containing the repository name of the charm (the org is assumed to be
   canonical)
2. Every charm in the bundle ahs a `_github_repo_branch` metadata field,
   containing the branch of the source code

> 

Author: Correct. Let me add this as well to the description there

```bash
./scripts/airgapped/get-all-images.sh releases/1.7/stable/kubeflow/bundle.yaml > images.txt
```

## Pull images to docker cache

We have a couple of scripts that are using `docker` commands to pull images,

> retag them and compress them in a  final `tar.gz` file. To save time on those

Author: Pushed a commit with this

> operations it's crucial to be able to first download/cache a list of images.

Author: Updated and removed the `To save time on those operations it's crucial to be able to first download/cache a list of images.`  part to mention that this script is used to just pull images, which is needed by following scripts.

```bash
python3 scripts/airgapped/save-images-to-cache.py images.txt
```

## Retag images to cache

In airgap environments users push their images in their own registries. So we'll
need to rename prefixes like `docker.io` to the server that users would use.

In our testing case our server is not DNS reachable from the machine. Instead
we directly access it with its IP and port `172.17.0.2:5000`

```bash
python3 scripts/airgapped/retag-images-to-cache.py images.txt
```

> Or if you'd like to use a different prefix, i.e. `registry.example.com`

Reviewer 2: Oh, I actually meant L48 and 49 starting with `In our testing case our server is not DNS reachable...`
This should be a very generic README.md that can help us run these set of scripts for any user or customer that requires an airgap installation, thus adding this example doesn't really add value to the readme. I agree that the example of how to use the python script with an example registry makes sense.

> Or if you'd like to use a different prefix, i.e. `registry.example.com`

Author: Pushed a commit that removed the `In our testing case...` line
